plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,318 @@
+
+const char *ffmpeg_hls_mux_getname(void *type)
+{
+	UNUSED_PARAMETER(type);
+	return obs_module_text("FFmpegHlsMuxer");
+}
+
+int hls_stream_dropped_frames(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	return stream->dropped_frames;
+}
+
+void ffmpeg_hls_mux_destroy(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+
+	if (stream) {
+		deactivate(stream, 0);
+
+		pthread_mutex_destroy(&stream->write_mutex);
+		os_sem_destroy(stream->write_sem);
+		os_event_destroy(stream->stop_event);
+
+		os_process_pipe_destroy(stream->pipe);
+		dstr_free(&stream->path);
+		dstr_free(&stream->printable_path);
+		dstr_free(&stream->stream_key);
+		dstr_free(&stream->muxer_settings);
+		bfree(data);
+	}
+}
+
+void *ffmpeg_hls_mux_create(obs_data_t *settings, obs_output_t *output)
+{
+	struct ffmpeg_muxer *stream = bzalloc(sizeof(*stream));
+	pthread_mutex_init_value(&stream->write_mutex);
+	stream->output = output;
+
+	/* init mutex, semaphore and event */
+	if (pthread_mutex_init(&stream->write_mutex, NULL) != 0)
+		goto fail;
+	if (os_event_init(&stream->stop_event, OS_EVENT_TYPE_AUTO) != 0)
+		goto fail;
+	if (os_sem_init(&stream->write_sem, 0) != 0)
+		goto fail;
+
+	UNUSED_PARAMETER(settings);
+	return stream;
+
+fail:
+	ffmpeg_hls_mux_destroy(stream);
+	return NULL;
+}
+
+static bool process_packet(struct ffmpeg_muxer *stream)
+{
+	struct encoder_packet packet;
+	int remaining;
+	bool ret = true;
+
+	pthread_mutex_lock(&stream->write_mutex);
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = stream->mux_packets.array[0];
+		da_erase(stream->mux_packets, 0);
+	}
+	pthread_mutex_unlock(&stream->write_mutex);
+
+	ret = write_packet(stream, &packet);
+	obs_encoder_packet_release(&packet);
+	return ret;
+}
+
+static void *write_thread(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+
+	while (os_sem_wait(stream->write_sem) == 0) {
+		if (os_event_try(stream->stop_event) == 0)
+			break;
+
+		bool ret = process_packet(stream);
+		if (!ret) {
+			int code = OBS_OUTPUT_ERROR;
+			obs_output_signal_stop(stream->output, code);
+			deactivate(stream, 0);
+			break;
+		}
+	}
+
+	os_atomic_set_bool(&stream->active, false);
+	return NULL;
+}
+
+bool ffmpeg_hls_mux_start(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	obs_service_t *service;
+	const char *path_str;
+	const char *stream_key;
+	struct dstr path = {0};
+	obs_encoder_t *vencoder;
+	obs_data_t *settings;
+	int keyint_sec;
+
+	if (!obs_output_can_begin_data_capture(stream->output, 0))
+		return false;
+	if (!obs_output_initialize_encoders(stream->output, 0))
+		return false;
+
+	service = obs_output_get_service(stream->output);
+	if (!service)
+		return false;
+	path_str = obs_service_get_url(service);
+	stream_key = obs_service_get_key(service);
+	dstr_copy(&stream->stream_key, stream_key);
+	dstr_copy(&path, path_str);
+	dstr_replace(&path, "{stream_key}", stream_key);
+	dstr_catf(&stream->muxer_settings,
```

Comment:
Does this result in muxer_settings getting longer and longer over multiple runs?

Reply:
I tested for that. Interestingly, it doesn't. However, to be sure, we could go back to calling dstr_init to 0 it out but free muxer_settings in deactivate, rather than destroy.